File: rust/src/protocol/v763/packets/play/clientbound/sound_effect.rs
//! Packet `sound_effect`. Hand-written.

use crate::codec::uuid_codec::{self as uuid_c, Uuid};
use crate::codec::{
    bitset, chat_component, identifier, nbt, position, slot, string_codec as string, varint,
    varlong, BytesReader, BytesWriter, Reader, Writer,
};
use crate::errors::ProtocolError;
use crate::protocol::v763::states::ConnectionState;
use crate::protocol::v763::ClientboundPacket;

pub const PACKET_ID: i32 = 0x62;

#[derive(Debug, Clone, PartialEq, Default)]
pub struct SoundEffect {
    pub sound_id: i32,
    pub custom_sound: Option<String>,
    pub custom_range: Option<f32>,
    pub sound_category: i32,
    pub x: i32,
    pub y: i32,
    pub z: i32,
    pub volume: f32,
    pub pitch: f32,
    pub seed: i64,
}
impl SoundEffect {
    pub fn decode(reader: &mut BytesReader<'_>) -> Result<Self, ProtocolError> {
        let sound_id = varint::read(reader)?;
        let (custom_sound, custom_range) = if sound_id == 0 {
            let cs = Some(identifier::read(reader)?);
            let p = reader.read_exact(1)?[0];
            let cr = match p {
                0 => None,
                1 => {
                    let b = reader.read_exact(4)?;
                    Some(f32::from_be_bytes([b[0], b[1], b[2], b[3]]))
                }
                o => return Err(ProtocolError::DecodeError(format!("range: {}", o))),
            };
            (cs, cr)
        } else {
            (None, None)
        };
        let sound_category = varint::read(reader)?;
        let b = reader.read_exact(12)?;
        let x = i32::from_be_bytes([b[0], b[1], b[2], b[3]]);
        let y = i32::from_be_bytes([b[4], b[5], b[6], b[7]]);
        let z = i32::from_be_bytes([b[8], b[9], b[10], b[11]]);
        let vb = reader.read_exact(8)?;
        let volume = f32::from_be_bytes([vb[0], vb[1], vb[2], vb[3]]);
        let pitch = f32::from_be_bytes([vb[4], vb[5], vb[6], vb[7]]);
        let sb = reader.read_exact(8)?;
        let seed = i64::from_be_bytes([sb[0], sb[1], sb[2], sb[3], sb[4], sb[5], sb[6], sb[7]]);
        Ok(Self {
            sound_id,
            custom_sound,
            custom_range,
            sound_category,
            x,
            y,
            z,
            volume,
            pitch,
            seed,
        })
    }
}
impl ClientboundPacket for SoundEffect {
    fn state(&self) -> ConnectionState {
        ConnectionState::Play
    }
    fn packet_id(&self) -> i32 {
        PACKET_ID
    }
    fn encode(&self, writer: &mut BytesWriter) -> Result<(), ProtocolError> {
        varint::write(self.sound_id, writer)?;
        if self.sound_id == 0 {
            identifier::write(self.custom_sound.as_deref().unwrap_or(""), writer)?;
            match self.custom_range {
                None => writer.write_all(&[0])?,
                Some(r) => {
                    writer.write_all(&[1])?;
                    writer.write_all(&r.to_be_bytes())?;
                }
            }
        }
        varint::write(self.sound_category, writer)?;
        writer.write_all(&self.x.to_be_bytes())?;
        writer.write_all(&self.y.to_be_bytes())?;
        writer.write_all(&self.z.to_be_bytes())?;
        writer.write_all(&self.volume.to_be_bytes())?;
        writer.write_all(&self.pitch.to_be_bytes())?;
        writer.write_all(&self.seed.to_be_bytes())?;
        Ok(())
    }
}

File: python/minecraft_bot/protocol/v763/packets/play/clientbound/sound_effect.py
"""Packet `sound_effect` (play/clientbound, id 0x62).

Plays a sound at world coordinates. The wire format begins with a
``sound`` selector — a VarInt sound ID; ``id == 0`` means the sound is
custom and the next bytes carry an ``Identifier`` plus an optional
``f32 range``.

Coordinates are in fixed-point eighth-blocks (i32 ``world * 8``).
"""

from __future__ import annotations

import struct
from dataclasses import dataclass

from minecraft_bot.codec import Reader, Writer, string, varint
from minecraft_bot.errors import ValueOutOfRange

PACKET_ID = 0x62


@dataclass(frozen=True, slots=True)
class SoundEffect:
    sound_id: int                # 0 = custom; 1+ = registry id
    custom_sound: str | None  # identifier; only if sound_id == 0
    custom_range: float | None  # f32, only when sound_id == 0 and range present
    sound_category: int          # varint enum
    x: int                       # i32 (world * 8)
    y: int                       # i32
    z: int                       # i32
    volume: float                # f32
    pitch: float                 # f32
    seed: int                    # i64


def decode(reader: Reader) -> SoundEffect:
    sid = varint.read(reader)
    if sid == 0:
        cs: str | None = string.read(reader)
        present = reader.read(1)[0]
        if present == 1:
            cr: float | None = struct.unpack(">f", reader.read(4))[0]
        elif present == 0:
            cr = None
        else:
            raise ValueOutOfRange("sound_effect.range.present", present)
    else:
        cs, cr = None, None
    cat = varint.read(reader)
    x, y, z = struct.unpack(">iii", reader.read(12))
    vol, pitch = struct.unpack(">ff", reader.read(8))
    seed, = struct.unpack(">q", reader.read(8))
    return SoundEffect(
        sound_id=sid, custom_sound=cs, custom_range=cr,
        sound_category=cat, x=x, y=y, z=z,
        volume=vol, pitch=pitch, seed=seed,
    )


def encode(packet: SoundEffect, writer: Writer) -> None:
    varint.write(packet.sound_id, writer)
    if packet.sound_id == 0:
        if packet.custom_sound is None:
            raise ValueOutOfRange("sound_effect.custom_sound", None)
        string.write(packet.custom_sound, writer)
        if packet.custom_range is None:
            writer.write(b"\x00")
        else:
            writer.write(b"\x01")
            writer.write(struct.pack(">f", packet.custom_range))
    varint.write(packet.sound_category, writer)
    writer.write(struct.pack(">iii", packet.x, packet.y, packet.z))
    writer.write(struct.pack(">ff", packet.volume, packet.pitch))
    writer.write(struct.pack(">q", packet.seed))
